rclcpp/src/rclcpp/parameter_events_subscriber.cpp

```diff
@@ -0,0 +1,308 @@
+  }
+}
+
+bool
+ParameterEventsSubscriber::should_unsubscribe_to_namespace(const std::string & node_namespace)
+{
+  auto it1 = std::find(event_namespaces_.begin(), event_namespaces_.end(), node_namespace);
+  if (it1 != event_namespaces_.end()) {
+    return false;
+  }
+
+  auto it2 = parameter_callbacks_.begin();
+  while (it2 != parameter_callbacks_.end()) {
+    if (split_path(it2->first.second).first == node_namespace) {
+      return false;
+    }
+    ++it2;
+  }
+  return true;
+}
+
+void
+ParameterEventsSubscriber::remove_parameter_callback(
+  const std::string & parameter_name,
+  const std::string & node_name)
+{
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  auto full_node_name = resolve_path(node_name);
+  parameter_callbacks_.erase({parameter_name, full_node_name});
+  if (should_unsubscribe_to_namespace(split_path(full_node_name).first)) {
+    RCLCPP_INFO(node_logging_->get_logger(), "Removing namespace event subscription");
+    remove_namespace_event_subscriber(split_path(full_node_name).first);
+  }
+}
+
+bool
+ParameterEventsSubscriber::get_parameter_from_event(
+  const rcl_interfaces::msg::ParameterEvent::SharedPtr event,
+  rclcpp::Parameter & parameter,
+  const std::string parameter_name,
+  const std::string node_name)
+{
+  if (event->node == node_name) {
+    rclcpp::ParameterEventsFilter filter(event, {parameter_name},
+      {rclcpp::ParameterEventsFilter::EventType::NEW,
+        rclcpp::ParameterEventsFilter::EventType::CHANGED});
+    if (!filter.get_events().empty()) {
+      const auto & events = filter.get_events();
+      auto param_msg = events.back().second;
+      parameter = rclcpp::Parameter::from_parameter_msg(*param_msg);
+      return true;
+    }
+  }
+  return false;
+}
+
+rclcpp::Parameter
+ParameterEventsSubscriber::get_parameter_from_event(
+  const rcl_interfaces::msg::ParameterEvent::SharedPtr event,
+  const std::string parameter_name,
+  const std::string node_name)
+{
+  rclcpp::Parameter p(parameter_name);
+  get_parameter_from_event(event, p, parameter_name, node_name);
+  return p;
+}
+
+void
+ParameterEventsSubscriber::event_callback(
+  const rcl_interfaces::msg::ParameterEvent::SharedPtr event)
+{
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  const std::string & node_name = event->node;
+  RCLCPP_DEBUG(node_logging_->get_logger(), "Parameter event received for node: %s",
+    node_name.c_str());
+
+  for (auto it = parameter_callbacks_.begin(); it != parameter_callbacks_.end(); ++it) {
+    rclcpp::Parameter p;
+    if (get_parameter_from_event(event, p, it->first.first, it->first.second)) {
+      for (auto cb = it->second.begin(); cb != it->second.end(); ++cb) {
+        auto shared_handle = cb->lock();
+        if (nullptr != shared_handle) {
+          shared_handle->callback(p);
+        } else {
+          cb = it->second.erase(cb);
+        }
+      }
+    }
+  }
+
+  if (event_callback_) {
+    event_callback_(event);
```

Comment:
Shouldn't call this one only when the namespace is in `event_namespaces_;`?

Reply:
I agree that makes more sense. 

Actually, I had a couple more thoughts I'm considering:
1) Apply the event callbacks per node as opposed to namespace
2) Support multiple event callbacks, either multiple on a same node, or multiple different nodes. This is similar to what I've done for the individual parameter callbacks